File: solutions/scala/118-implement-queue-using-stacks.scala
/**
 * Problem 118: Implement Queue using Stacks (LeetCode 232)
 * Difficulty: Easy
 * Language: Scala
 */
case class FQueue[A](inStack: List[A] = Nil, outStack: List[A] = Nil) {
  def enqueue(x: A): FQueue[A] = copy(inStack = x :: inStack)
  def transfer: FQueue[A] =
    if (outStack.nonEmpty) this
    else FQueue(Nil, inStack.reverse)
  def dequeue: (A, FQueue[A]) = {
    val t = transfer
    (t.outStack.head, t.copy(outStack = t.outStack.tail))
  }
  def peek: A = transfer.outStack.head
}

File: solutions/typescript/118-implement-queue-using-stacks.ts
/**
 * Problem 118: Implement Queue using Stacks (LeetCode 232)
 * Difficulty: Easy
 * Language: TypeScript
 */
type Queue<T> = { inStack: T[]; outStack: T[] };
const emptyQueue = <T>(): Queue<T> => ({ inStack: [], outStack: [] });
const enqueue = <T>(q: Queue<T>, x: T): Queue<T> =>
  ({ ...q, inStack: [...q.inStack, x] });
const transfer = <T>(q: Queue<T>): Queue<T> =>
  q.outStack.length > 0 ? q
    : { inStack: [], outStack: [...q.inStack].reverse() };
const dequeue = <T>(q: Queue<T>): [T, Queue<T>] => {
  const tq = transfer(q);
  return [tq.outStack[0], { ...tq, outStack: tq.outStack.slice(1) }];
};
const peek = <T>(q: Queue<T>): T => transfer(q).outStack[0];
